src/server/tiered_storage.cc

```diff
@@ -390,6 +402,45 @@ void TieredStorage::FinishIoRequest(int io_res, InflightWriteRequest* req) {
   VLOG_IF(2, num_active_requests_ == 0) << "Finished active requests";
 }
 
+PrimeIterator TieredStorage::Load(DbIndex db_index, PrimeIterator it, string_view key) {
+  PrimeValue* entry = &it->second;
+  CHECK(entry->IsExternal());
+  DCHECK_EQ(entry->ObjType(), OBJ_STRING);
+  auto [offset, size] = entry->GetExternalSlice();
+  string res;
+  res.resize(size);
+  auto ec = Read(offset, size, res.data());
+  CHECK(!ec) << "TBD";
+
+  // Read will preempt, check if iterator still points to our entry
+  PrimeTable* pt = db_slice_.GetTables(db_index).first;
+  if (!it.is_occupied() || it->first != key) {
+    it = pt->Find(key);
+    if (it.is_done()) {
+      // Entry was remove from db while reading from disk. (background expire task)
+      return it;
+    }
+    entry = &it->second;
+  }
+
+  if (!entry->IsExternal()) {
+    // Because 2 reads can happen at the same time, then if the other read
+    // already loaded the data from disk to memeory we dont need to do anything now just return.
+    // TODO we can register to reads with multiple callbacks so if there is already a callback
+    // reading the data from disk we will not run read twice.
+    return it;
+  }
+
+  auto stats = db_slice_.MutableStats(db_index);
+  Free(it, stats);
+  entry->SetString(res);
+
+  size_t heap_size = entry->MallocUsed();
+  stats->obj_memory_usage += heap_size;
+  stats->AddTypeMemoryUsage(entry->ObjType(), heap_size);
```

Comment:
It looks like we always modify `obj_memory_usage` when we call `AddTypeMemoryUsage()`, so maybe have `AddTypeMemoryUsage()` also update `obj_memory_usage`?